opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -2539,6 +2555,127 @@ namespace Opm {
 
 
 
+
+
+    template<typename TypeTag>
+    void
+    BlackoilWellModel<TypeTag>::
+    runWellPIScaling(const int timeStepIdx, DeferredLogger& local_deferredLogger)
+    {
+        if (this->last_run_wellpi_.has_value() && (*this->last_run_wellpi_ == timeStepIdx)) {
```

Comment:
let me confirm here. If we use the keyword `WELPI`, do we do this for each following report steps?

The manual is not very clear to me, I got the impression it is a one-time thing when reading it. 

do we handle the situation that connections change (added new connections or connection tranmissibility changes with new input?)

Reply:
Static `WELPI`s&mdash;i.e., those that are listed as separate keywords and not triggered as the result of a dynamic action (`ACTIONX`)&mdash;act as follows

1. At the start of the report step for which `WELPI` is specified, compute the dynamic PI value for the pertinent well's "preferred" phase
2. Use this dynamic PI value to compute a scaling factor as `WPI(2) / dynamic_PI`
3. Apply that scaling factor to the transmissibility factor of every connection in the well, whether they are open or not.  Moreover, this scaling should persist until the the transmissibility factor is explicitly reset in COMPDAT.

The last operation is handled by [`Schedule::applyWellProdIndexScaling()`](https://github.com/OPM/opm-common/blob/0102d736acead804ee843852f62f32cf5d85333a/src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp#L1545).  This PR is about handling steps 1 and 2.

> do we handle the situation that connections change (added new connections or connection tranmissibility changes with new input)?

Yes.  That's built into `Schedule::applyWellProdIndexScaling()`.

> Under these circumstances, the well PI will not be the one specified in WELPI.

That is correct.  `WELPI` is an instantaneous scaling based on the flow conditions of a single phase.  The dynamic PI values will typically drift away from the requested value fairly soon after the scaling has been applied, but will start from that new value.